client/src/lib/result.js

```diff
@@ -40,3 +47,68 @@ export function paddedAttemptResults(result, numberOfAttempts) {
   const attemptResults = result.attempts.map((attempt) => attempt.result);
   return padSkipped(attemptResults, numberOfAttempts);
 }
+
+// Events sorted by best don't need forecast view.
+// Fewest moves is currently unsupported
+// Only final rounds are supported
+export function forecastViewEnabled(round) {
+  return round.format.sortBy != "best" &&
+    round.competitionEvent.event.id != "333fm" &&
+    round.name === "Final";
+}
+
+/**
+ * return the results object with additional properties:
+ * projectedAverage: projected final average based on current results
+ * forFirst: time needed to overtake first place
+ * forThird: time needed to overtake third place
+ */
+export function resultsForView(results, format, forecastView) {
+  if (results.length == 0 || !forecastView) return results;
+  var resultsForView = results.map((result) => {
+    return {
+      ...result,
+      projectedAverage: computeProjectedAverage(result, format),
+    };
+  });
+
+  if (resultsForView[0].projectedAverage == SKIPPED_VALUE) {
+    // First place has no results. Do nothing.
+    return resultsForView;
+  }
+
+  for (let result of resultsForView) {
+    result.advancingQuestionable = false;
+  }
+
+  // Forecast view only supported for final rounds - advancing count hardcoded as 3
+  const advancingCount = 3;
+
+  // Sort based on projection with tiebreakers on single
+  resultsForView = orderBy(resultsForView, [
+    (result) => toMonotonic(result.projectedAverage),
+    (result) => toMonotonic(result.best),
+  ]);
+  resultsForView[0].ranking = 1;
+  var prevResult = resultsForView[0];
+  for (let i = 0; i < resultsForView.length; i++) {
+    let currentResult = resultsForView[i];
+    if (isSkipped(currentResult.projectedAverage)) {
+      break;
+    }
+    if (currentResult.projectedAverage === prevResult.projectedAverage &&
+      currentResult.best === prevResult.best) {
+      // Rankings tie
+      currentResult.ranking = prevResult.ranking;
+    } else {
+      currentResult.ranking = i + 1;
+    }
+    if (currentResult.ranking <= advancingCount) {
+      // "advancing" field is handled correctly by default results query
+      currentResult.advancingQuestionable = true;
+    }
```

Comment:
I believe the logic we want is this:

```suggestion
    const isClinched = currentResult.advancing && currentResult.advancingQuestionable;
    
    // A clinched result must still be clinched in the projected ranking,
    // so we keep advancing state as is
    if (!isClinched) {
      if (currentResult.ranking <= advancingCount) {
        currentResult.advancing = true;
        currentResult.advancingQuestionable = true;
      } else {
        currentResult.advancing = false;
        currentResult.advancingQuestionable = false;
      }
    }
```

Let's make sure there is a test for the clinched scenario.

Reply:
I think I'm confused on how advancing and advancingQuestionable work. It looks like advancingQuestionable doesn't do anything if advancing is set to true?

For this view, I was thinking that we would have the advancing competitors be defined by the query results from the server. Then the new rankings based on projectedAverage would change the advancingQuestionable field since these are based on projected times and so we don't want to indicate they are clinched (advancing = true).

Reply:
Actually it looks like for clinched results `advancing=true` but `advancingQuestionable=false`? It looks like the formatting for the questionable property also override so that if both are true then the result has the light highlighting.

Wouldn't clinched then be `currentResult.advancing && !currentResult.advancingQuestionable`

Reply:
I updated it so that `advancingQuestionable` isn't altered if `advancing` is set (since questionable overrides default). Otherwise I think it should work as is?

Reply:
Ok I cleaned this up and think I have it in a good spot. Just want to confirm a typo in your suggestion that what it should be is `const isClinched = currentResult.advancing && !currentResult.advancingQuestionable;`. Then the rest makes sense

Reply:
Gahh, yes I meant `const isClinched = currentResult.advancing && !currentResult.advancingQuestionable`. Sorry for the confusion :)